### PR TITLE
Add pyi stubs for `view_*` and `ViewerModel`

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
See #2458 for context: our autogenerated `napari.view_*` and `viewer.add_*` methods present problems for both IDE autocompletion, and cognitive burden for new contributors.

This PR doesn't address the second issue, but it does add `.pyi` stubs for `napari.view_layers` and `napari.components.viewer_model`.  Each of those modules can be run as modules (e.g. `python -m napari.view_layers`) to generate the respective `.pyi` files.  Those pyi files _are_ checked into source here, so that does create the issue of needing to keep those files in sync with the source.  We can either do it manually when necessary, or happy to implement an auto-run task if someone has thoughts on how/when.

**edit:** I added an item to the `makefile`, so you can run `make typestubs`

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by tlambert03** [comment]
I added docstrings to the `pyi` files too... but can remove them.  That part is strictly for IDEs that provide docs on hover, etc...

File: napari/utils/misc.py [hunk 1]
```diff
@@ -441,3 +449,55 @@ def dir_hash(path: Union[str, Path], include_paths=True, ignore_hidden=True):
...
+
+def _generate_cls_stubs(cls, output_file, imports: Sequence[str] = ()) -> str:
+    """Generate pyi-format type stubs for a class."""
+    import textwrap
+
+    from ..utils.misc import get_subclass_methods
+
+    bases = ", ".join(f'{b.__module__}.{b.__name__}' for b in cls.__bases__)
+
+    pyi = '# THIS FILE IS AUTOGENERATED.  DO NOT EDIT\n'
```
Comment (kne42) [inline napari/utils/misc.py:462]: could you also add a comment saying that the code to generate it can be found here?

File: napari/utils/misc.py [hunk 2]
```diff
@@ -441,3 +449,55 @@ def dir_hash(path: Union[str, Path], include_paths=True, ignore_hidden=True):
                 _hash.update(''.join(fparts).encode())
 
     return _hash.hexdigest()
+
+
+def _generate_cls_stubs(cls, output_file, imports: Sequence[str] = ()) -> str:
+    """Generate pyi-format type stubs for a class."""
+    import textwrap
+
+    from ..utils.misc import get_subclass_methods
```
Comment (kne42) [inline napari/utils/misc.py:458]: I'm a bit confused here, aren't both of these functions in `napari.utils.misc`?

File: napari/view_layers.py [hunk 3]
```diff
@@ -146,15 +149,42 @@ def _generate_view_function(layer_string: str, method_name: str = None):
...
-    'vectors',
-    'tracks',
-):
+for _layer in layers.NAMES:
     _generate_view_function(_layer)
 
 _generate_view_function('path', 'open')
+
+
+def _generate_stubs(output=__file__.replace(".py", ".pyi")):
```
Comment (kne42) [inline napari/view_layers.py:158]: how come `_generate_cls_stubs` is in `napari.utils.misc` and this is not?

File: napari/view_layers.py [hunk 4]
```diff
@@ -146,15 +149,42 @@ def _generate_view_function(layer_string: str, method_name: str = None):
...
+    """Generat type stubs for view_* functions declared in this file."""
```
Comment (kne42) [inline napari/view_layers.py]:
```suggestion
    """Generate type stubs for view_* functions declared in this file."""
```

File: napari/view_layers.py [hunk 5]
```diff
@@ -146,15 +149,42 @@ def _generate_view_function(layer_string: str, method_name: str = None):
...
+    import textwrap
+
+    pyi = '# THIS FILE IS AUTOGENERATED.  DO NOT EDIT\n'
```
Comment (kne42) [inline napari/view_layers.py:162]: could you also please reference the source of this file in a comment?

**Review by kne42 — COMMENTED** [review]
Awesome PR @tlambert03! I wonder if we can also just build the typestubs on install like what @jni was suggesting in #2458? I suppose that would require that we make `black` a requirement though...

> I added docstrings to the pyi files too... but can remove them. That part is strictly for IDEs that provide docs on hover, etc...

Nice! Have you checked if this is working properly?

File: napari/utils/misc.py (hunk 2 above)
Comment (tlambert03) [reply napari/utils/misc.py:458]: hah.  they weren't at first :)  thanks for catching!

File: napari/view_layers.py (hunk 3 above)
Comment (tlambert03) [reply napari/view_layers.py:158]: this one was a bit hard to make generic...  but I should have tried harder :)  I can make it `generate_module_stubs` and have it accept the names that should be added

**Comment by tlambert03** [comment]
> Awesome PR @tlambert03! I wonder if we can also just build the typestubs on install like what @jni was suggesting in #2458? I suppose that would require that we make black a requirement though...

yeah we can definitely do that.  Could just run `make typestubs` in our build (and just install `black` and `isort` in the CI env)

> > I added docstrings to the pyi files too... but can remove them. That part is strictly for IDEs that provide docs on hover, etc...

> Nice! Have you checked if this is working properly?

yep!

Untitled 2

**Comment by goanpeca** [comment]
Awesome @tlambert03. How about generating the stubs as part of the pre-commit ?

**Comment by tlambert03** [comment]
> How about generating the stubs as part of the pre-commit ?

I think pre-commit might be a bit too frequent (and it will add a little more lag to every commit, which is annoying)... but I have a local modification that will run it during `setup.py`, so we could actually consider not checking it all into source, (and running `pip install -e .` would also update stubs it in a dev environment if need be).  Thoughts?

**Comment by tlambert03** [comment]
see #2675 instead

**Pull request closed without merging**